File: mobile/lib/services/nai_stream.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:msgpack_dart/msgpack_dart.dart' as msgpack;

const _maxFrameBytes = 128 * 1024 * 1024;

class NaiGenerationPreview {
  final Uint8List image;
  final double progress;
  final int currentStep;
  final int totalSteps;
  final int sampleIndex;
  final bool finalImage;

  const NaiGenerationPreview({
    required this.image,
    required this.progress,
    required this.currentStep,
    required this.totalSteps,
    required this.sampleIndex,
    required this.finalImage,
  });
}

class NaiStreamResult {
  final List<Uint8List> images;
  final Uint8List? archive;

  const NaiStreamResult({this.images = const [], this.archive});
}

class NaiStreamException implements Exception {
  final String message;
  final bool previewStarted;

  const NaiStreamException(this.message, {this.previewStarted = false});

  @override
  String toString() => message;
}

class _Frame {
  final String eventType;
  final int sampleIndex;
  final int? stepIndex;
  final Uint8List? image;
  final String? error;

  const _Frame({
    required this.eventType,
    required this.sampleIndex,
    this.stepIndex,
    this.image,
    this.error,
  });
}

Map<String, dynamic>? _record(Object? value) {
  if (value is! Map) return null;
  return value.map((key, value) => MapEntry(key.toString(), value));
}

int? _optionalInteger(Object? value) =>
    value is num ? value.round() : int.tryParse(value?.toString() ?? '');

bool _looksLikeBase64(String value) {
  final compact = value.replaceAll(RegExp(r'\s+'), '');
  return compact.length >= 4 &&
      compact.length % 4 == 0 &&
      RegExp(r'^[A-Za-z0-9+/]*={0,2}$').hasMatch(compact);
}

Uint8List? _decodeImage(Object? value) {
  if (value is Uint8List) return Uint8List.fromList(value);
  if (value is ByteBuffer) return value.asUint8List();
  if (value is List && value.every((entry) => entry is int)) {
    return Uint8List.fromList(value.cast<int>());
  }
  if (value is String && value.isNotEmpty) {
    final compact = value
        .replaceFirst(
            RegExp(r'^data:image/[^;]+;base64,', caseSensitive: false), '')
        .replaceAll(RegExp(r'\s+'), '');
    if (!_looksLikeBase64(compact)) return null;
    try {
      return base64Decode(compact);
    } catch (_) {
      return null;
    }
  }
  return null;
}

_Frame _frameFromRecord(Map<String, dynamic> message,
    [String fallbackEventType = '']) {
  final nested = _record(message['payload']) ?? _record(message['data']);
  final source = nested != null &&
          message['image'] == null &&
          message['event_type'] == null
      ? {...message, ...nested}
      : message;
  final eventType = (source['event_type'] ??
          source['eventType'] ??
          source['type'] ??
          fallbackEventType.ifEmpty('intermediate'))
      .toString();
  final error = source['message'] ?? source['error'];
  return _Frame(
    eventType: eventType,
    sampleIndex: _optionalInteger(source['samp_ix'] ??
            source['sampleIndex'] ??
            source['sample_index']) ??
        0,
    stepIndex: _optionalInteger(
        source['step_ix'] ?? source['stepIndex'] ?? source['step_index']),
    image: _decodeImage(source['image'] ??
        source['data'] ??
        source['image_data'] ??
        source['imageData']),
    error: eventType == 'error' || source['error'] != null
        ? (error ?? 'Stream generation failed').toString()
        : null,
  );
}

_Frame? _decodeMessagePack(Uint8List bytes) {
  final value = msgpack.deserialize(bytes, copyBinaryData: true);
  final message = _record(value);
  return message == null ? null : _frameFromRecord(message);
}

class _MessagePackFrameDecoder {
  Uint8List _pending = Uint8List(0);

  List<_Frame> push(List<int> chunk) {
    if (chunk.isEmpty) return const [];
    final next = Uint8List(_pending.length + chunk.length)
      ..setRange(0, _pending.length, _pending)
      ..setRange(_pending.length, _pending.length + chunk.length, chunk);
    _pending = next;
    final output = <_Frame>[];
    var offset = 0;
    while (_pending.length - offset >= 4) {
      final data = ByteData.sublistView(_pending, offset, offset + 4);
      final length = data.getUint32(0, Endian.big);
      if (length <= 0 || length > _maxFrameBytes) {
        throw NaiStreamException(
            'Invalid NovelAI stream frame length: $length');
      }
      if (_pending.length - offset < 4 + length) break;
      final frame = _decodeMessagePack(
          Uint8List.sublistView(_pending, offset + 4, offset + 4 + length));
      if (frame != null) output.add(frame);
      offset += 4 + length;
    }
    if (offset > 0) _pending = Uint8List.sublistView(_pending, offset);
    return output;
  }
}

class _SseFrameDecoder {
  final List<int> _pending = [];
  String _eventType = '';
  final List<String> _dataLines = [];

  List<_Frame> _dispatch() {
    if (_dataLines.isEmpty) {
      _eventType = '';
      return const [];
    }
    final value = _dataLines.join('\n').trim();
    final eventType = _eventType;
    _eventType = '';
    _dataLines.clear();
    if (value.isEmpty || value == '[DONE]') return const [];
    try {
      final parsed = jsonDecode(value);
      final record = _record(parsed);
      if (record != null) return [_frameFromRecord(record, eventType)];
      if (parsed is String) return _decodeSseValue(eventType, parsed);
    } catch (_) {
      // Proxies may send base64 MessagePack or a direct image instead of JSON.
    }
    return _decodeSseValue(eventType, value);
  }

  List<_Frame> _decodeSseValue(String eventType, String value) {
    if (!_looksLikeBase64(value)) {
      return eventType == 'error'
          ? [_Frame(eventType: eventType, sampleIndex: 0, error: value)]
          : const [];
    }
    final bytes = base64Decode(value.replaceAll(RegExp(r'\s+'), ''));
    try {
      final decoded = _decodeMessagePack(bytes);
      if (decoded != null) {
        return [
          _Frame(
            eventType:
                decoded.eventType == 'intermediate' && eventType.isNotEmpty
                    ? eventType
                    : decoded.eventType,
            sampleIndex: decoded.sampleIndex,
            stepIndex: decoded.stepIndex,
            image: decoded.image,
            error: decoded.error,
          )
        ];
      }
    } catch (_) {
      // A valid base64 payload can also be the preview image itself.
    }
    return [
      _Frame(
        eventType: eventType.ifEmpty('intermediate'),
        sampleIndex: 0,
        image: bytes,
      )
    ];
  }

  List<_Frame> push(List<int> chunk) {
    _pending.addAll(chunk);
    final output = <_Frame>[];
    while (true) {
      final newline = _pending.indexOf(0x0a);
      if (newline < 0) break;
      final raw = _pending.sublist(0, newline);
      _pending.removeRange(0, newline + 1);
      if (raw.isNotEmpty && raw.last == 0x0d) raw.removeLast();
      final line = utf8.decode(raw, allowMalformed: true);
      if (line.isEmpty) {
        output.addAll(_dispatch());
      } else if (!line.startsWith(':')) {
        final colon = line.indexOf(':');
        final field = colon < 0 ? line : line.substring(0, colon);
        var value = colon < 0 ? '' : line.substring(colon + 1);
        if (value.startsWith(' ')) value = value.substring(1);
        if (field == 'event') _eventType = value.trim();
        if (field == 'data') _dataLines.add(value);
      }
    }
    return output;
  }

  List<_Frame> finish() {
    if (_pending.isNotEmpty) {
      var raw = List<int>.from(_pending);
      _pending.clear();
      if (raw.isNotEmpty && raw.last == 0x0d) raw.removeLast();
      final line = utf8.decode(raw, allowMalformed: true);
      final colon = line.indexOf(':');
      final field = colon < 0 ? line : line.substring(0, colon);
      var value = colon < 0 ? '' : line.substring(colon + 1);
      if (value.startsWith(' ')) value = value.substring(1);
      if (field == 'event') _eventType = value.trim();
      if (field == 'data') _dataLines.add(value);
    }
    return _dispatch();
  }
}

enum _StreamMode { unknown, messagePack, sse, zip }

Future<NaiStreamResult> consumeNaiGenerationStream(
  Stream<List<int>> stream, {
  required int totalSteps,
  required void Function(NaiGenerationPreview preview) onPreview,
  String contentType = '',
}) async {
  final messagePack = _MessagePackFrameDecoder();
  final sse = _SseFrameDecoder();
  final finals = <int, Uint8List>{};
  var mode = contentType.toLowerCase().contains('text/event-stream')
      ? _StreamMode.sse
      : _StreamMode.unknown;
  final prefix = <int>[];
  final zip = BytesBuilder(copy: false);
  var previewStarted = false;
  var lastPreviewAt = DateTime.fromMillisecondsSinceEpoch(0);

  void consumeFrames(List<_Frame> frames) {
    for (final frame in frames) {
      if (frame.error != null) {
        throw NaiStreamException(frame.error!, previewStarted: previewStarted);
      }
      final image = frame.image;
      if (image == null || image.isEmpty) continue;
      previewStarted = true;
      final currentStep = (frame.stepIndex ?? 0) + 1;
      final finalImage = frame.eventType == 'final';
      if (finalImage) finals[frame.sampleIndex] = image;
      final now = DateTime.now();
      if (finalImage ||
          now.difference(lastPreviewAt).inMilliseconds >= 110 ||
          currentStep >= totalSteps) {
        lastPreviewAt = now;
        onPreview(NaiGenerationPreview(
          image: image,
          progress: finalImage
              ? 1
              : (currentStep / maxOf(1, totalSteps)).clamp(0, .99).toDouble(),
          currentStep: finalImage ? totalSteps : currentStep,
          totalSteps: totalSteps,
          sampleIndex: frame.sampleIndex,
          finalImage: finalImage,
        ));
      }
    }
  }

  try {
    await for (final chunk in stream) {
      if (mode == _StreamMode.unknown) {
        prefix.addAll(chunk);
        if (prefix.length < 4) continue;
        final textual = utf8
            .decode(prefix.take(32).toList(), allowMalformed: true)
            .replaceFirst('\ufeff', '')
            .trimLeft();
        final length = ByteData.sublistView(Uint8List.fromList(prefix), 0, 4)
            .getUint32(0, Endian.big);
        mode = prefix[0] == 0x50 && prefix[1] == 0x4b
            ? _StreamMode.zip
            : RegExp(r'^(?:event|data|id|retry)\s*:|^:').hasMatch(textual)
                ? _StreamMode.sse
                : length > 0 && length <= _maxFrameBytes
                    ? _StreamMode.messagePack
                    : _StreamMode.sse;
        if (mode == _StreamMode.zip) {
          zip.add(prefix);
        } else if (mode == _StreamMode.sse) {
          consumeFrames(sse.push(prefix));
        } else {
          consumeFrames(messagePack.push(prefix));
        }
        prefix.clear();
        continue;
      }
      if (mode == _StreamMode.zip) {
        zip.add(chunk);
      } else if (mode == _StreamMode.sse) {
        consumeFrames(sse.push(chunk));
      } else {
        consumeFrames(messagePack.push(chunk));
      }
    }
    if (mode == _StreamMode.zip) {
      return NaiStreamResult(archive: zip.takeBytes());
    }
    if (mode == _StreamMode.sse) consumeFrames(sse.finish());
    if (finals.isEmpty) {
      throw NaiStreamException(
        '流式生成结束，但没有收到最终图片。为避免重复扣费，未自动重发请求。',
        previewStarted: previewStarted,
      );
    }
    final entries = finals.entries.toList()
      ..sort((left, right) => left.key.compareTo(right.key));
    return NaiStreamResult(
        images: entries.map((entry) => entry.value).toList());
  } on NaiStreamException {
    rethrow;
  } catch (error) {
    throw NaiStreamException(error.toString(), previewStarted: previewStarted);
  }
}

int maxOf(int left, int right) => left > right ? left : right;

extension _StringFallback on String {
  String ifEmpty(String fallback) => isEmpty ? fallback : this;
}

File: electron/ipc/nai-stream.ts
import { decode } from "@msgpack/msgpack";

export interface NaiStreamFrame {
  eventType: "intermediate" | "final" | "error" | string;
  sampleIndex: number;
  stepIndex?: number;
  image?: Buffer;
  error?: string;
}

const MAX_FRAME_BYTES = 128 * 1024 * 1024;

function asRecord(value: unknown): Record<string, unknown> | null {
  if (value instanceof Map) {
    return Object.fromEntries(
      Array.from(value.entries(), ([key, entry]) => [String(key), entry]),
    );
  }
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return null;
}

function optionalInteger(value: unknown): number | undefined {
  const parsed = typeof value === "number" ? value : Number(value);
  return Number.isFinite(parsed) ? Math.trunc(parsed) : undefined;
}

function looksLikeBase64(value: string): boolean {
  const compact = value.replace(/\s+/g, "");
  return compact.length >= 4
    && compact.length % 4 === 0
    && /^[A-Za-z0-9+/]*={0,2}$/.test(compact);
}

function decodeImage(value: unknown): Buffer | undefined {
  if (Buffer.isBuffer(value)) return value;
  if (value instanceof Uint8Array) {
    return Buffer.from(value.buffer, value.byteOffset, value.byteLength);
  }
  if (value instanceof ArrayBuffer) return Buffer.from(value);
  if (Array.isArray(value) && value.every((entry) => Number.isInteger(entry))) {
    return Buffer.from(value as number[]);
  }
  if (typeof value === "string" && value.length > 0) {
    try {
      const compact = value.replace(/^data:image\/[^;]+;base64,/i, "").replace(/\s+/g, "");
      if (!looksLikeBase64(compact)) return undefined;
      return Buffer.from(compact, "base64");
    } catch {
      return undefined;
    }
  }
  return undefined;
}

function frameFromRecord(
  message: Record<string, unknown>,
  fallbackEventType?: string,
): NaiStreamFrame {
  const nested = asRecord(message.payload) ?? asRecord(message.data);
  const source = nested && !message.image && !message.event_type
    ? { ...message, ...nested }
    : message;
  const eventType = String(
    source.event_type
      ?? source.eventType
      ?? source.type
      ?? fallbackEventType
      ?? "intermediate",
  );
  const errorValue = source.message ?? source.error;
  return {
    eventType,
    sampleIndex: optionalInteger(source.samp_ix ?? source.sampleIndex ?? source.sample_index) ?? 0,
    stepIndex: optionalInteger(source.step_ix ?? source.stepIndex ?? source.step_index),
    image: decodeImage(source.image ?? source.data ?? source.image_data ?? source.imageData),
    error:
      eventType === "error" || source.error !== undefined
        ? String(errorValue ?? "Stream generation failed")
        : undefined,
  };
}

export function decodeNaiStreamFrame(bytes: Uint8Array): NaiStreamFrame | null {
  const message = asRecord(decode(bytes));
  if (!message) return null;
  return frameFromRecord(message);
}

function decodeSseData(eventType: string, value: string): NaiStreamFrame | null {
  const data = value.trim();
  if (!data || data === "[DONE]") return null;

  try {
    const parsed = JSON.parse(data) as unknown;
    const record = asRecord(parsed);
    if (record) return frameFromRecord(record, eventType || undefined);
    if (typeof parsed === "string") return decodeSseData(eventType, parsed);
  } catch {
    // A number of NovelAI-compatible gateways wrap the MessagePack payload or
    // the preview image directly in base64 instead of JSON. Try those forms
    // below without turning malformed text into arbitrary image bytes.
  }

  if (!looksLikeBase64(data)) {
    if (eventType === "error") {
      return { eventType, sampleIndex: 0, error: data };
    }
    return null;
  }
  const bytes = Buffer.from(data.replace(/\s+/g, ""), "base64");
  try {
    const decoded = decodeNaiStreamFrame(bytes);
    if (decoded) {
      if ((!decoded.eventType || decoded.eventType === "intermediate") && eventType) {
        decoded.eventType = eventType;
      }
      return decoded;
    }
  } catch {
    // Direct image payload. Image validation is intentionally lightweight here;
    // the renderer will decode the same bytes and the final archive parser still
    // validates the generated file before saving it.
  }
  return {
    eventType: eventType || "intermediate",
    sampleIndex: 0,
    image: bytes,
  };
}

/** Incremental Server-Sent Events decoder used by NovelAI-compatible proxy
 * endpoints. The official endpoint currently uses framed MessagePack, while
 * some gateways return the same events as `event:` / `data:` records. */
export class NaiSseFrameDecoder {
  private pending = "";
  private eventType = "";
  private dataLines: string[] = [];
  private readonly textDecoder = new TextDecoder("utf-8");

  private dispatch(): NaiStreamFrame[] {
    if (!this.dataLines.length) {
      this.eventType = "";
      return [];
    }
    const frame = decodeSseData(this.eventType, this.dataLines.join("\n"));
    this.eventType = "";
    this.dataLines = [];
    return frame ? [frame] : [];
  }

  private consumeLines(final = false): NaiStreamFrame[] {
    const frames: NaiStreamFrame[] = [];
    let newline = this.pending.indexOf("\n");
    while (newline >= 0) {
      let line = this.pending.slice(0, newline);
      this.pending = this.pending.slice(newline + 1);
      if (line.endsWith("\r")) line = line.slice(0, -1);
      if (!line) {
        frames.push(...this.dispatch());
      } else if (!line.startsWith(":")) {
        const colon = line.indexOf(":");
        const field = colon < 0 ? line : line.slice(0, colon);
        let value = colon < 0 ? "" : line.slice(colon + 1);
        if (value.startsWith(" ")) value = value.slice(1);
        if (field === "event") this.eventType = value.trim();
        else if (field === "data") this.dataLines.push(value);
      }
      newline = this.pending.indexOf("\n");
    }
    if (final) {
      if (this.pending) {
        const line = this.pending.endsWith("\r") ? this.pending.slice(0, -1) : this.pending;
        const colon = line.indexOf(":");
        const field = colon < 0 ? line : line.slice(0, colon);
        let value = colon < 0 ? "" : line.slice(colon + 1);
        if (value.startsWith(" ")) value = value.slice(1);
        if (field === "event") this.eventType = value.trim();
        else if (field === "data") this.dataLines.push(value);
      }
      this.pending = "";
      frames.push(...this.dispatch());
    }
    return frames;
  }

  push(chunk: Uint8Array): NaiStreamFrame[] {
    if (chunk.byteLength === 0) return [];
    this.pending += this.textDecoder.decode(chunk, { stream: true });
    return this.consumeLines();
  }

  finish(): NaiStreamFrame[] {
    this.pending += this.textDecoder.decode();
    return this.consumeLines(true);
  }
}

/**
 * Incrementally decodes NovelAI's stream protocol:
 * a four-byte big-endian message length followed by one MessagePack object.
 */
export class NaiStreamFrameDecoder {
  private pending = Buffer.alloc(0);

  push(chunk: Uint8Array): NaiStreamFrame[] {
    if (chunk.byteLength === 0) return [];
    const incoming = Buffer.from(chunk.buffer, chunk.byteOffset, chunk.byteLength);
    this.pending = this.pending.length
      ? Buffer.concat([this.pending, incoming])
      : Buffer.from(incoming);
    const frames: NaiStreamFrame[] = [];
    while (this.pending.length >= 4) {
      const length = this.pending.readUInt32BE(0);
      if (length <= 0 || length > MAX_FRAME_BYTES) {
        throw new Error(`Invalid NovelAI stream frame length: ${length}`);
      }
      if (this.pending.length < 4 + length) break;
      const frame = decodeNaiStreamFrame(this.pending.subarray(4, 4 + length));
      this.pending = this.pending.subarray(4 + length);
      if (frame) frames.push(frame);
    }
    return frames;
  }

  remainingBytes(): Buffer {
    return Buffer.from(this.pending);
  }
}

export function frameNaiStreamMessage(message: Uint8Array): Buffer {
  const header = Buffer.allocUnsafe(4);
  header.writeUInt32BE(message.byteLength, 0);
  return Buffer.concat([
    header,
    Buffer.from(message.buffer, message.byteOffset, message.byteLength),
  ]);
}
